### ISSUE TITLE
Scaled FP4 GEMM accuracy issue

### ISSUE THREAD

**Issue by TheTinyTeddy** [mit-han-lab/fouroversix#9]
```

import torch
torch.manual_seed(123)

import transformer_engine.pytorch as te
from transformer_engine.common.recipe import Format, NVFP4BlockScaling
from fouroversix import quantize_to_fp4, fp4_matmul, MatmulBackend, AdaptiveBlockScalingRule, RoundStyle
from fouroversix.quantize.reference import dequantize_from_fp4, from_blocked

def create_nvfp4_linear(in_features, out_features, bias=None):
    linear = te.Linear(
        in_features, 
        out_features, 
        bias=bias,
        params_dtype=torch.bfloat16, 
        device='cuda'
    )
    return linear


def test_nvfp4_linear(linear, x):
    # Setup NVFP4 recipe
    fp4_format = Format.E2M1
    fp4_recipe = NVFP4BlockScaling(
        fp4_format=fp4_format,
        disable_rht=True,
        disable_stochastic_rounding=True,
        disable_2d_quantization=False,
    )
    
    # Forward pass with NVFP4 autocast
    with te.autocast(enabled=True, recipe=fp4_recipe):
        output = linear(x)
    
    return output

def manual_fp4_matmul_reference(x, weight):
    # Ensure bf16
    if x.dtype != torch.bfloat16:
        x = x.to(torch.bfloat16)
    if weight.dtype != torch.bfloat16:
        weight = weight.to(torch.bfloat16)
    
    # Quantize to FP4
    x_e2m1, x_sf, x_normconst = quantize_to_fp4(
        x,
        scale_rule=AdaptiveBlockScalingRule.always_6,
    )
    
    weight_contiguous = weight.contiguous()
    weight_e2m1, weight_sf, weight_normconst = quantize_to_fp4(
        weight_contiguous,
        scale_rule=AdaptiveBlockScalingRule.always_6,
        block_scale_2d=True,
    )
    
    # FP4 matmul
    output = fp4_matmul(
        a_e2m1=x_e2m1,
        a_sf=x_sf,
        a_normconst=x_normconst,
        b_e2m1=weight_e2m1,
        b_sf=weight_sf,
        b_normconst=weight_normconst,
        backend=MatmulBackend.cutlass,
        out_shape=(x.shape[0], weight.shape[0])
    )
    
    return output

def main():
    
    # Test configuration
    in_features = 768
    out_features = 1024
    batch_size = 4
    seq_len = 128

    
    # Create inputs
    x = torch.randn(batch_size, seq_len, in_features, dtype=torch.bfloat16, device='cuda')

    
    # Create a single weight tensor to ensure both layers have identical weights
    shared_weight = torch.randn(out_features, in_features, dtype=torch.bfloat16, device='cuda') * 0.02
    shared_bias = None
    
    # Create NVFP4 linear and assign shared weight
    nvfp4_linear = create_nvfp4_linear(in_features, out_features, bias=True)
    with torch.no_grad():
        nvfp4_linear.weight.copy_(shared_weight)
    
    

    # NVFP4 forward
    nvfp4_output = test_nvfp4_linear(nvfp4_linear, x.view(-1, in_features))
    manual_output = manual_fp4_matmul_reference(x.view(-1, in_features), nvfp4_linear.weight)
    torch_output = x.view(-1, in_features)@shared_weight.T

    TEnvfp4_diff=torch.dist(nvfp4_output,torch_output)
    fouroversix_diff=torch.dist(manual_output,torch_output)

    print('TEnvfp4_diff: ',TEnvfp4_diff.item(), '     fouroversix_diff: ',fouroversix_diff.item())

    
if __name__ == "__main__":
    for _ in range(10):
        main()
```

This codes produce:
TEnvfp4_diff:  58.75      fouroversix_diff:  147.0
TEnvfp4_diff:  58.75      fouroversix_diff:  145.0
TEnvfp4_diff:  58.75      fouroversix_diff:  145.0
TEnvfp4_diff:  58.75      fouroversix_diff:  142.0
TEnvfp4_diff:  58.75      fouroversix_diff:  144.0
TEnvfp4_diff:  58.75      fouroversix_diff:  153.0
TEnvfp4_diff:  58.75      fouroversix_diff:  143.0
TEnvfp4_diff:  58.75      fouroversix_diff:  149.0
TEnvfp4_diff:  58.75      fouroversix_diff:  143.0
TEnvfp4_diff:  58.75      fouroversix_diff:  142.0

which indicates accuracy issue with the current fouroversix implementation

**Comment by TheTinyTeddy** [comment]
Interesting. In theory, even in the event of register spilling, the compiler should still ensure the functional correctness.. 

Or like you said this could be a pure Triton bug before the low-level compiler stage
